### PR TITLE
Fix promotion in cov and mean to handle integer and rational matrices

### LINKED ISSUE #8080: Promotion in cov
See discussion at:  http://stackoverflow.com/questions/25052037/cova-mean-0-gives-inexacterror

I did not find any reference to it when searching the issues.  Apologies in advance if the issue has been filed already.

I just got a similar error from generic_scale!, reached via `in cor at statistics.jl:439.`
The versioninfo for my machine is:

```
Julia Version 0.3.0
Commit 7681878 (2014-08-20 20:43 UTC)
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-2720QM CPU @ 2.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes #8080

### CODE REVIEW

File: base/statistics.jl [hunk 1]
```diff
@@ -335,12 +335,18 @@ unscaled_covzm(x::AbstractMatrix, y::AbstractMatrix, vardim::Int) =
 # covzm (with centered data)
 
 covzm(x::AbstractVector; corrected::Bool=true) = unscaled_covzm(x) / (_length(x) - Int(corrected))
-covzm(x::AbstractMatrix, vardim::Int=1; corrected::Bool=true) =
-    scale!(unscaled_covzm(x, vardim), inv(size(x,vardim) - Int(corrected)))
+function covzm(x::AbstractMatrix, vardim::Int=1; corrected::Bool=true)
+    C = unscaled_covzm(x, vardim)
+    T = typeof(first(C) / 1)
```
Comment (tkelman) [inline base/statistics.jl]: can unscaled_covzm ever return heterogeneously-typed arrays?

Comment (andreasnoack) [inline base/statistics.jl]: It can. Are you worried about `first` not being representative of the elements in `C`?

Comment (tkelman) [inline base/statistics.jl]: it should usually be a subtype of the element type, but other elements might be wider for abstractly typed arrays, right?

Comment (andreasnoack) [inline base/statistics.jl]: The problem is that the element type has to be fairly concrete before we can instantiate it but I've just realized that in this case here, it should be possible to use `promote_type(typeof(first(C) / 1), eltype(x))`. I'll update.

Comment (tkelman) [inline base/statistics.jl]: `eltype(x)` or `eltype(C)`? will `unscaled_covzm` ever result in a different eltype?

Comment (andreasnoack) [inline base/statistics.jl]: Probably rare. I can only think of `Bool` entries and there is wouldn't change `T` but I think you are right that I should use `C` and not `x`

**Pull request merged**